### PR TITLE
Ppc record replay

### PR DESCRIPTION
**Pull request by wangray**
TODO: Pass CI, write tests
Also add stats?

### CODE REVIEW

File: panda/src/rr/rr_print.c [hunk 1]
```diff
@@ -43,7 +44,7 @@ volatile sig_atomic_t rr_end_record_requested = 0;
 volatile sig_atomic_t rr_end_replay_requested = 0;
 char * rr_requested_name = NULL;
 
-// write this program point to this file 
+// write this program point to this file
```
Comment (phulin) [inline panda/src/rr/rr_print.c]: can you change your editor to not make a bunch of unrelated whitespace fixes?

File: target/ppc/excp_helper.c [hunk 2]
```diff
@@ -753,6 +758,14 @@ static void ppc_hw_interrupt(CPUPPCState *env)
...
 #endif
     /* External reset */
+
+		int pending_interrupts;
+		RR_DO_RECORD_OR_REPLAY(
+				pending_interrupts = env->pending_interrupts,
+					rr_record_pending_interrupts(RR_CALLSITE_CPU_PENDING_INTERRUPTS, pending_interrupts);,
+				rr_replay_pending_interrupts((uint32_t*)&env->pending_interrupts),
+				 RR_CALLSITE_CPU_PENDING_INTERRUPTS);
+
```
Comment (phulin) [inline target/ppc/excp_helper.c]: fix indentation

File: panda/src/rr/rr_log.c [hunk 3]
```diff
@@ -436,6 +448,33 @@ void rr_record_interrupt_request(RR_callsite_id call_site,
...
+
+    memset(item, 0, sizeof(RR_log_entry));
+    item->header.kind = RR_PENDING_INTERRUPTS;
+    item->header.callsite_loc = call_site;
+    item->header.prog_point = rr_prog_point();
+
+    item->variant.pending_interrupts = pending_int;
+
+    rr_write_item();
+    rr_nondet_log->current_state = RR_INTERRUPT_PENDING;
```
Comment (phulin) [inline panda/src/rr/rr_log.c]: Still not sure why you added this rr_set_state stuff instead of just recording changes to the pending_interrupts field?

File: panda/scripts/diverge.py [hunk 4]
```diff
@@ -230,6 +235,17 @@ def condition_instr(self, break_arg, op, instr):
...
 
+    def set_breakpoint_commands(self, break_num):
+        self.gdb("commands", break_num, expect_prompt = ">")
+        # self.gdb("p/u cpus->tqh_first->rr_guest_instr_count", expect_prompt = ">")
+        self.gdb("call target_disas(stdout, cpu, tb->pc, tb->size, 0)", expect_prompt = ">")
+        self.gdb("end")
+
+    def display_commands(self):
+        self.display("cpus->tqh_first->rr_guest_instr_count")
+        self.gdb("set $env = ((CPUPPCState*) cpus->tqh_first->env_ptr)")
```
Comment (phulin) [inline panda/scripts/diverge.py]: make arch-independent, and also no need for the set command. i assume this stuff is just here for convenience?

File: panda/src/common.c [hunk 5]
```diff
@@ -10,7 +10,7 @@
 target_ulong panda_current_pc(CPUState *cpu) {
     target_ulong pc, cs_base;
     uint32_t flags;
-    CPUArchState *env = cpu->env_ptr;
```
Comment (phulin) [inline panda/src/common.c]: no need for unrelated whitespace changes

File: panda/scripts/diverge.py (hunk 4 above)
Comment (wangray) [inline panda/scripts/diverge.py]: yep. just for my personal use. I can remove

File: panda/src/rr/rr_log.c (hunk 3 above)
Comment (wangray) [inline panda/src/rr/rr_log.c]: I started out with that approach. Will try to see if i can make it work now.

File: build.sh [hunk 6]
```diff
@@ -39,6 +39,8 @@ else
   fi
 fi
 
+LLVM_BIT=""
```
Comment (phulin) [inline build.sh]: what's going on here?

Comment (wangray) [inline build.sh]: Something tim told me to add b/c i was having llvm problems? I'll remove since it's no longer relevant.

Comment (phulin) [inline build.sh]: you can run with --disable-llvm on the command line i think

File: target/ppc/excp_helper.c [hunk 7]
```diff
@@ -753,6 +759,14 @@ static void ppc_hw_interrupt(CPUPPCState *env)
                   cs->interrupt_request, (int)msr_me, (int)msr_ee);
 #endif
     /* External reset */
+
```
Comment (phulin) [inline target/ppc/excp_helper.c]: fix indentation

File: target/ppc/excp_helper.c [hunk 8]
```diff
@@ -739,6 +744,7 @@ void ppc_cpu_do_interrupt(CPUState *cs)
     PowerPCCPU *cpu = POWERPC_CPU(cs);
     CPUPPCState *env = &cpu->env;
 
+	//TODO: Need to record this...
```
Comment (phulin) [inline target/ppc/excp_helper.c]: you are recording this!

File: target/ppc/excp_helper.c [hunk 9]
```diff
@@ -24,10 +24,15 @@
...
 
+#ifdef CONFIG_SOFTMMU
+#include "panda/rr/rr_log.h"
+#endif
+
 //#define DEBUG_OP
 //#define DEBUG_SOFTWARE_TLB
 //#define DEBUG_EXCEPTIONS
 
+
```
Comment (phulin) [inline target/ppc/excp_helper.c]: try to avoid unnecessary changes, they cause merge conflicts :)

File: panda/src/rr/rr_log.c (hunk 3 above)
Comment (phulin) [inline panda/src/rr/rr_log.c]: Can you still do this?

File: panda/scripts/run_on_32bitlinux.py [hunk 10]
```diff
@@ -117,21 +133,24 @@ def EXIT_USAGE():
     if not os.path.exists(install_dir):
         os.mkdir(install_dir)
 
-    qcow = join(dot_dir, "wheezy_panda2.qcow2")
+    qcow = join(dot_dir, "i386_wheezy.qcow")
```
Comment (phulin) [inline panda/scripts/run_on_32bitlinux.py]: this isn't generic

File: panda/scripts/run_on_32bitlinux.py [hunk 11]
```diff
@@ -117,21 +133,24 @@ def EXIT_USAGE():
...
     if not os.path.isfile(qcow):
         print "\nYou need a qcow. Downloading from moyix. Thanks moyix!\n"
-        sp.check_call(["wget", "http://panda.moyix.net/~moyix/wheezy_panda2.qcow2", "-O", qcow])
+        # sp.check_call(["wget", "http://panda.moyix.net/~moyix/wheezy_panda2.qcow2", "-O", qcow])
```
Comment (phulin) [inline panda/scripts/run_on_32bitlinux.py]: neither is this

File: cpu-exec.c [hunk 12]
```diff
@@ -498,7 +497,19 @@ static inline bool cpu_handle_exception(CPUState *cpu, int *ret)
             cpu->exception_index = -1;
             return true;
 #else
+            //TODO: Record/replay exceptions...
```
Comment (phulin) [inline cpu-exec.c]: you did it!

File: .gitignore [hunk 13]
```diff
@@ -1,3 +1,6 @@
+/.idea/*
+.DS_Store
+Cmakelists.txt
```
Comment (phulin) [inline .gitignore]: What's the deal with this stuff? If it's necessary you should put it with the rest of the panda-specific stuff.

Comment (wangray) [inline .gitignore]: Artifacts of OSX and my editor...

Comment (wangray) [inline .gitignore]: I'll remove it

File: panda/scripts/run_on_32bitlinux.py (hunk 10 above)
Comment (wangray) [inline panda/scripts/run_on_32bitlinux.py]: Yea, i'll just make it a command line arg.

File: panda/src/rr/rr_log.c (hunk 3 above)
Comment (wangray) [inline panda/src/rr/rr_log.c]: Wilco.
